docs/source/model_design/programming_vivarium/debugging/index.rst

```diff
@@ -0,0 +1,190 @@
+Let's continue with the example above, where we got an error 
+in a pandas checking function. 
+
+The python debugger places you at the end of the stack trace - 
+so we are currently sitting in a pandas error message function. 
+We can check this by typing :code:`l.` or :code:`ll` which prints 
+the code around the current line (l.), or a longer version of this 
+print out (ll). This will let you see what the code surrounding the 
+error is. Likely in this case, it will just show you that pandas found 
+mismatched lengths and threw the error message, which isn't too helpful. 
+
+Next, you can move up or down the stack trace with :code:`u` or 
+:code:`d` respectively. Since we start at the bottom of the stack 
+trace, we will move up first. By entering :code:`u`, we move to the 
+pandas checking function, then the pandas multiplication function, and finally 
+to the line in our :code:`loader.py` function. 
+
+Note: it can 
+often take several times entering :code:`u` to get to a useful 
+point in the code. Don't be surprised if you need to move up a lot!
+
+Now in the debugger, we are "located" in the :code:`loader.py` function 
+that started this error. Since the debugger is "in" this line of code, 
+the objects at that point in the code are able to be called. For example, 
+we can print or manipulate the dataframes being worked with. The code 
+for this in the python debugger is the same as used in 
+a Jupyter notebook. By printing the dataframes being used, you 
+might discover that one of the dataframes is empty - and the 
+issue is more clear! 
+
+From there, we can look at the code generating the empty dataframe. 
+Maybe you see the error quickly. But if not, we'll probably 
+want to add a breakpoint. 
+
+To quit the python debugger, enter :code:`quit`. 
+
+Adding Breakpoints
+------------------
+
+So, we have identified the issue is an empty dataframe. Let's keep 
+debugging! The datframe in question is generated 
+in a function not in our stack trace. This means we'll have trouble 
+getting to it with the python debugger. We'll need to set a breakpoint. 
+
+A breakpoint is a line of code :code:`breakpoint()` that "breaks" 
+the code. So when this line is read, the model will drop you into a 
+python debugger at that point in the code. So in our example, we might 
+set a breakpoint in the function which creates the empty dataframe. 
+
+Once in the python debugger, we can use a similar process to the above. 
+Start by printing the dataframe. If it's empty, the issue is earlier 
+in the code. If the dataframe is still correct, the issue is later. 
+Either use up or step (:code:`s`) to move up or "step" down to the next 
+line. In this manner, you can navigate around the code in the debugger and 
+find which line is the issue. From there, you can investigate the line 
+and fix the bug. 
+
+If you need to use multiple breakpoints, you can also use the 
+command :code:`c` for continue, to move until the next breakpoint. 
+
+Using iPython
```

Comment:
I have a strong feeling based on this section that you might prefer to use VSCode for this purpose -- happy to demo if you'd like!

Reply:
Hey Zeb - maybe separately from this PR we can chat about IPython alternatives - the VSCode and having Jupyter not hog your terminal. I think I'm still going to merge this now but maybe next docs session we can run through these options and see what we want as the default in the docs.